pkg/xds/server/callbacks/dataplane_sync_callbacks.go

```diff
@@ -0,0 +1,230 @@
+	if metadata == nil {
+		return errors.New("metadata in xDS Node cannot be nil")
+	}
+	if metadata.Resource != nil {
+		if core_model.MetaToResourceKey(metadata.Resource.GetMeta()) != dpKey {
+			return fmt.Errorf("proxyId %s does not match proxy resource %s", dpKey, metadata.Resource.GetMeta())
+		}
+	}
+	l := dataplaneSyncLog.WithValues("dpKey", dpKey, "streamID", streamID)
+	t.Lock()
+	if t.activeStreams[streamID].proxyInfo != nil {
+		return nil // We fast return if we already know that this streamID is tracking a specific dataplane
+	}
+	pInfo := t.proxyInfos[dpKey]
+	if pInfo == nil {
+		ctx, cancel := context.WithCancel(context.Background())
+		done := make(chan struct{})
+		pInfo = &dpInfo{
+			dpKey:            dpKey,
+			lifecycleManager: t.dpLifecycleManagerFactory.New(dpKey),
+			ctx:              ctx,
+			done:             done,
+			meta:             atomic.Pointer[core_xds.DataplaneMetadata]{},
+			cancelFunc:       cancel,
+		}
+		t.proxyInfos[dpKey] = pInfo
+	}
+	t.activeStreams[streamID].proxyInfo = pInfo
+	ctx := t.activeStreams[streamID].ctx
+	pInfo.Lock()
+	defer pInfo.Unlock()
+	t.Unlock()
+	pInfo.streams = append(pInfo.streams, streamID)
+	pInfo.meta.Store(metadata)
+	if len(pInfo.streams) == 1 {
+		err := pInfo.lifecycleManager.Register(l, ctx, metadata)
+		if err != nil {
+			return err
+		}
+
+		l.V(1).Info("starting watchdog")
+		go func() {
+			defer close(pInfo.done)
+			l.V(1).Info("watchdog started")
+			if t.dpWatchdogFactory != nil {
+				t.dpWatchdogFactory.New(dpKey, pInfo.meta.Load).Start(pInfo.ctx)
+			}
+		}()
+	} else {
+		l.V(1).Info("watchdog was started by a previous stream, not starting a new one", "activeStreams", pInfo.streams)
+	}
+
+	return nil
+}
+
+func (t *dataplaneSyncCallbacks) OnStreamClosed(streamID core_xds.StreamID) {
+	t.Lock()
+	l := dataplaneSyncLog.WithValues("streamID", streamID)
+	l.V(1).Info("stream closed")
+	sInfo := t.activeStreams[streamID]
+	if sInfo == nil { // Should never happen as it means a stream was closed without being opened
+		l.Error(errors.New("stream closed without being opened"), "stream closed without being opened")
+		t.Unlock()
+		return
+	}
+	if sInfo.proxyInfo == nil { // Never associated with a dataplane, no need to care
+		delete(t.activeStreams, streamID)
+		t.Unlock()
+		return
+	}
+	pInfo := sInfo.proxyInfo
+	l = dataplaneSyncLog.WithValues("dpKey", pInfo.dpKey)
+	pInfo.Lock()
+	t.Unlock()
+	streamIdx := slices.Index(pInfo.streams, streamID)
+	if streamIdx == -1 {
+		l.Info("streamID not found in the list of streams", "streams", pInfo.streams)
+		pInfo.Unlock()
+		return
+	}
+	if streamIdx != len(pInfo.streams)-1 {
+		l.V(1).Info("dpKey is associated with a newer stream", "streams", pInfo.streams)
+		pInfo.streams = append(pInfo.streams[:streamIdx], pInfo.streams[streamIdx+1:]...) // Remove streamID from the list
+		pInfo.Unlock()
+		return // We are not the last stream, we don't care
+	}
+	if sInfo.proxyInfo.cancelFunc != nil {
```

Comment:
since you released the `t.Unlock()` you can't use `sInfo` as it could be accessed for example from the `OnStreamRequest`:
```go
t.activeStreams[streamID].proxyInfo = pInfo
```

Reply:
Here `sInfo` is a copy to the value in the map, right? So this statement should still work even the proxyInfo was re-assigned in the map from the method `OnStreamRequest`.